sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -975,6 +939,73 @@ object PushPredicateThroughAggregate extends Rule[LogicalPlan] with PredicateHel
       } else {
         filter
       }
+
+    case filter @ Filter(condition, u: Union) =>
```

Comment:
"non-determinstic" means the condition has internal state and may change every time it processes an input row. Before push down, we have only one stateful condition that processes all output rows of `Union`. But if we push it down, then each child of `Union` will have a stateful condition, I think it's different than before.